### PR TITLE
use services.Config.NewService/Engine

### PR DESCRIPTION
**Pull request by jmank88**
https://smartcontract-it.atlassian.net/browse/BCF-3337

This PR upgrades a few `Services` from using a low-level `StateMachine`/`WaitGroup`/`StopChan` to a new API provided by common.

### Requires Dependencies
- https://github.com/smartcontractkit/chainlink-common/pull/204

### CODE REVIEW

**Review by krehermann — DISMISSED** [review]

**Review by EasterTheBunny — DISMISSED** [review]

**Review by EasterTheBunny — DISMISSED** [review]

File: core/chains/evm/monitor/balance.go [hunk 1]
```diff
@@ -53,59 +54,44 @@ var _ BalanceMonitor = (*balanceMonitor)(nil)
...
+		ethKeyStore: ethKeyStore,
+		ethBalances: make(map[gethCommon.Address]*assets.Eth),
 	}
-	bm.sleeperTask = utils.NewSleeperTask(&worker{bm: bm})
+	bm.Service, bm.eng = services.Config{
+		Name:  "BalanceMonitor",
+		Start: bm.start,
+		Close: bm.close,
+	}.NewServiceEngine(lggr)
+	bm.sleeperTask = utils.NewSleeperTask(&worker{bm: bm}) //TODO convert to sub-service?
```
Comment (ilija42) [inline core/chains/evm/monitor/balance.go]: Should this be a ticket?

File: core/chains/evm/monitor/balance.go [hunk 2]
```diff
@@ -198,12 +184,12 @@ func (w *worker) checkAccountBalance(ctx context.Context, address gethCommon.Add
```
Comment (ilija42) [inline core/chains/evm/monitor/balance.go:182]: Is this  `const ethFetchTimeout = 15 * time.Second` a relic before layer 2s and alt evm chains? Its out of scope but maybe we should raise this somewhere?

Comment (ilija42) [reply core/chains/evm/monitor/balance.go:182]: Sleeper task in common also times in 15seconds by default

File: core/chains/evm/monitor/balance.go [hunk 3]
```diff
@@ -53,59 +54,44 @@ var _ BalanceMonitor = (*balanceMonitor)(nil)
...
+	ok := bm.sleeperTask.IfStarted(func() {
 		bm.checkBalance(head)
 	})
 	if !ok {
-		bm.logger.Debugw("BalanceMonitor: ignoring OnNewLongestChain call, balance monitor is not started", "state", bm.State())
+		bm.eng.Debugw("BalanceMonitor: ignoring OnNewLongestChain call, balance monitor is not started", "state", bm.sleeperTask.State())
 	}
 }
 
 func (bm *balanceMonitor) checkBalance(head *evmtypes.Head) {
```
Comment (ilija42) [inline core/chains/evm/monitor/balance.go]: also out of scope but may deserve a TODO. Looks like this should be called checkBalances and shouldn't accept any args

File: core/recovery/recover.go [hunk 4]
```diff
@@ -3,38 +3,38 @@ package recovery
...
 
-			panic(err)
-		}
-	}()
-	fn()
+		panic(err)
+	})
 }
 
 func WrapRecover(lggr logger.Logger, fn func()) {
```
Comment (ilija42) [inline core/recovery/recover.go:20]: What is the point of WrapRecover? I see all callers giving it a func for onPanic arg which then gets ignored

File: core/services/fluxmonitorv2/flux_monitor.go [hunk 5]
```diff
@@ -136,9 +133,13 @@ func NewFluxMonitor(
...
 		chProcessLogs: make(chan struct{}, 1),
-		chStop:        make(services.StopChan),
-		waitOnStop:    make(chan struct{}),
 	}
+	fm.Service, fm.eng = services.Config{
+		Name:  "FluxMonitor",
+		Start: fm.start,
+		Close: fm.close,
+	}.NewServiceEngine(lggr)
+	fm.logger = logger.Sugared(fm.eng)
```
Comment (ilija42) [inline core/services/fluxmonitorv2/flux_monitor.go:142]: Is there a reason why we still need a logger outside of the one in the engine?

File: core/services/nurse.go [hunk 6]
```diff
@@ -66,85 +66,63 @@ const (
...
+		for reason, checkFunc := range n.checks {
+			if unwell, meta := checkFunc(); unwell {
+				n.GatherVitals(ctx, reason, meta)
+				break
+			}
+		}
 	})
-}
 
-func (n *Nurse) Close() error {
```
Comment (ilija42) [inline core/services/nurse.go:140]: Maybe keep Close for the closing log

File: core/services/nurse.go [hunk 7]
```diff
@@ -66,85 +66,63 @@ const (
...
+	return n
 }
 
-func (n *Nurse) Start() error {
-	return n.StartOnce("Nurse", func() error {
-		// This must be set *once*, and it must occur as early as possible
-		if n.cfg.MemProfileRate() != runtime.MemProfileRate {
-			runtime.MemProfileRate = n.cfg.BlockProfileRate()
-		}
+func (n *Nurse) start(ctx context.Context) error {
```
Comment (ilija42) [inline core/services/nurse.go]: ctx is unused here

Comment (ilija42) [reply core/services/nurse.go]: oh, I guess this is to satisfy the interface, maybe `start(_ context.Context) error {` then

**Review by ilija42 — CHANGES_REQUESTED** [review]

File: core/recovery/recover.go (hunk 4 above)
Comment (jmank88) [reply core/recovery/recover.go:20]: This variant does not have a custom onPanic arg, so it just logs critical.

File: core/chains/evm/monitor/balance.go (hunk 2 above)
Comment (jmank88) [reply core/chains/evm/monitor/balance.go:182]: Seems like it, but I also don't see any reason for this timeout to be related to block times at all. It is not as if the call would block until a block is produced. :confused:

File: core/services/nurse.go (hunk 6 above)
Comment (jmank88) [reply core/services/nurse.go:140]: The engine will already log

File: core/services/fluxmonitorv2/flux_monitor.go (hunk 5 above)
Comment (jmank88) [reply core/services/fluxmonitorv2/flux_monitor.go:142]: This type used sugared logger methods, so this is just creating it once.

File: core/chains/evm/monitor/balance.go (hunk 1 above)
Comment (jmank88) [reply core/chains/evm/monitor/balance.go]: I think I can just drop it

File: core/chains/evm/monitor/balance.go (hunk 2 above)
Comment (jmank88) [reply core/chains/evm/monitor/balance.go:182]: https://smartcontract-it.atlassian.net/browse/BCI-3948

**Review by ilija42 — APPROVED** [review]

**Review by EasterTheBunny — APPROVED** [review]

**Pull request merged**